ee/modules/160-multitenancy-manager/hooks/project.go

```diff
@@ -0,0 +1,192 @@
+			FilterFunc: filterProjectTypesForUpdateProjects,
+		},
+	},
+}, handleProjects)
+
+type projectSnapshot struct {
+	Name            string
+	Template        map[string]*apiext.JSON
+	ProjectTypeName string
+	Conditions      []v1alpha1.Condition
+}
+
+func filterProjects(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	pt := &v1alpha1.Project{}
+	if err := sdk.FromUnstructured(obj, pt); err != nil {
+		return nil, err
+	}
+
+	return projectSnapshot{
+		Name:            pt.Name,
+		ProjectTypeName: pt.Spec.ProjectTypeName,
+		Template:        pt.Spec.Template,
+		Conditions:      pt.Status.Conditions,
+	}, nil
+}
+
+func filterProjectTypesForUpdateProjects(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	pt := &v1alpha1.ProjectType{}
+	if err := sdk.FromUnstructured(obj, pt); err != nil {
+		return nil, err
+	}
+
+	return pt.Spec, nil
+}
+
+type projectValues struct {
+	Params          map[string]interface{} `json:"params"`
+	ProjectTypeName string                 `json:"projectTypeName"`
+	ProjectName     string                 `json:"projectName"`
+}
+
+func handleProjects(input *go_hook.HookInput) error {
+	projectSnapshots := input.Snapshots[projectsQueue]
+	if len(projectSnapshots) < 1 {
+		return nil
+	}
+
+	values := make([]projectValues, 0, len(projectSnapshots))
+	for _, projectSnap := range projectSnapshots {
+		if projectSnap == nil {
+			continue
+		}
+
+		project, ok := projectSnap.(projectSnapshot)
+		if !ok {
+			input.LogEntry.Errorf("can't convert snapshot to 'projectSnapshot': %v", project)
+			continue
+		}
+
+		errStatusProject := setErrorProjectWrap(input.PatchCollector, project.Name, project.Conditions)
+
+		ptSpecValues, ok := input.Values.GetOk(internal.ModuleValuePath(internal.PTValuesPath, project.ProjectTypeName))
+		if !ok {
+			errMsg := fmt.Sprintf("Can't find valid ProjectType '%s' for Project", project.ProjectTypeName)
+			errStatusProject(errMsg)
+			continue
+		}
+
+		ptValues := ptSpecValues.Value()
+		ptValuesMap, ok := ptValues.(map[string]interface{})
+		if !ok {
+			errMsg := fmt.Sprintf("can't convert '%s' ProjectType values to map[string]interface: %T", project.ProjectTypeName, ptValues)
+			errStatusProject(errMsg)
+			continue
+		}
+
+		schema, err := internal.LoadOpenAPISchema(ptValuesMap["openAPI"])
+		if err != nil {
+			errMsg := fmt.Sprintf("can't load '%s' ProjectType OpenAPI schema: %v", project.ProjectTypeName, err)
+			errStatusProject(errMsg)
+			continue
+		}
+
+		projectTemplateValues, err := mapStringJSONToMapStringInterface(project.Template)
+		if err != nil {
+			errMsg := fmt.Sprintf("can't convert map[string]*apiext.JSON to map[string]interface{} for template values: %v", err)
+			errStatusProject(errMsg)
+			continue
+		}
+
+		if schema.AdditionalProperties == nil {
+			schema.AdditionalProperties = &spec.SchemaOrBool{Allows: false}
+		}
+		if err := validate.AgainstSchema(schema, projectTemplateValues, strfmt.Default); err != nil {
```

Comment:
It doesn't work for me in some cases. I created a project type with the following OpenAPI:
```yaml
    openAPI:
      limits:
        properties:
          cpu:
            oneOf:
            - format: int
              type: number
            - type: string
            pattern: ^[0-9]+m?$
          memory:
            oneOf:
            - format: int
              type: number
            - type: string
            pattern: ^[0-9]+(\.[0-9]+)?(E|P|T|G|M|k|Ei|Pi|Ti|Gi|Mi|Ki)?$
        type: object
      requests:
        properties:
          cpu:
            oneOf:
            - format: int
              type: number
            - type: string
            pattern: ^[0-9]+m?$
          memory:
            oneOf:
            - format: int
              type: number
            - type: string
            pattern: ^[0-9]+(\.[0-9]+)?(E|P|T|G|M|k|Ei|Pi|Ti|Gi|Mi|Ki)?$
```
Then I created a project with the following template:
```yaml
  template:
    limits:
      cpu: 5
      memory: 5Gi
    requests:
      cpu: 5
      memory: 5Gi
      storage: 1Gi
```

There is no storage option in openapi, but it is allowed in the template. We should probably use the same hack as for deckhouse - disable additional properties on all scheme levels.